### DISCUSSION TITLE
Incorrect Kalman estimation

### DISCUSSION THREAD

**Discussion by Noel700** [bitcraze/discussions#846] [Q&A] [answered]
Hi,

I'm using the CF2.1 with Lighthouse and Multiranger. My Goal is to start at a position and autonomous fly to a target Position. For this I'm calculating the Trajectory to Target myself. I'm not using the High Level Commander, cause i wanna implement obstacle avoidance during that motion and go_to() won't return until Target is reached. My Trajectory seems fine. Starting at e.g. x: -1.2 and y: 0.3 wanting to x: 1 and y: 0.
My CF completes it flight and lands at a position. I return my Kalman Estimates right before the landing which are giving me values like x: -0.4 and y: -0.17. So that way of from Target. But i see that the real Position is very close to my Target. Going into the client, manually connecting and reading the values from the Lighthouse Tab i get like x: 0.9 and y: 0.04. So it really is close and the Kalman Estimate that return is wrong. I'm ignoring the z Value for the moment cause it's not changing during the flight but it might also get a bad Estimation if it does.
Also going into the Client I quite permanently get "Kalman prediction rate off". Using the newest firmware release.
Any Idea why is that and how i can get the correct Value which is shown in the Lighthouse Tab? I can also provide some Code if needed but i just get the Values via LogConfig and 'kalman.stateX' so on.

Thanks in advance
Noel

**Comment by knmcguire** [comment]
Hi!

That is a bit odd but perhaps this has to do with the geometry estimation. The values of the lighthouse tab are actually the values that come from the Kalman estimates. 

Perhaps mark some locations you know are true on the floor, and just manually (not flying) move your crazyflie to those locations. While doing this, look at the cfclient at your kalman.stateX in the parameter tab or [stateEstimate.x](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/logs/#stateestimate) (this is what the lighthouse tab is using).

**Reply by Noel700** [reply]
Thanks for your suggestions. I marked and measured some points on the floor. I calculated the geometry again as well. The raw precision seems perfect. I can perfectly move and fly a meter or two. The values in the Lighthouse Tab are perfect as well. But I still get a Console log which states: ESTKALMAN: WARNING: Kalman prediction rate low (98)
Theres another Msg that shows quite often: ESTKALMAN: State out of bounds, resetting

Also it seems like this sometimes forces my Cf into a reboot. Because sometimes it just reboots during a flight and shows an Error with radiolink. Searching for the same Error, I found this post: https://forum.bitcraze.io/viewtopic.php?t=4797 but didn't got any good Information out of it

**Reply by Noel700** [reply]
To add a bit more Information. I'm using a 4 Base station Lighthouse Setup. Taking a look into the Lighthouse tab I see, that some Basestation Receiving is blinking between green and red. So there sometimes isn't a constant receiving. Maybe that's causing my Problem? Is there a best way to turn the Basestations? Because it feels like they are positioned very good and the Drone should be able to receive constantly

**Reply by knmcguire** [reply, marked as answer]
I think that 4 base-stations receiving on one lighthouse is currently it's limit in the newest firmware. You could try to just turn 2 basestations off and see if it improves at least at that side of the basestations.

Rebooting is a different story... does it only reboot, or does it reboot and that the red led flikkers? You can then connect to the crazyflie when it is in that state to get the error message out, as it would save on where in the firmware it went wrong.

**Reply by Noel700** [reply]
Thanks, I tried it with only 2 Basestations and it seems to work very well now. The Kalman Estimation that i get in my Python Program is accurate now. The warnings went away and as far as I see it right now, the drone didn't restart.
About the restarts: I opened another thread about that. It was a radiolink error. As I understood it the Chip was so busy with Kalman Estimation and Lighthouse that it ran out of resources with 4 Basestations

**Reply by knmcguire** [reply]
It's strange.. it should be able to handle 4 basestations. We can make an issue out of this. What was your flight space (location of the basestations?)

**Reply by Noel700** [reply]
Sorry for the late reply. I was thinking so as well. Anyway it would be great if you can investigate this a bit more.
I got my Stations set up in a rectangle. It's approximitly 6m x 4x. They are in about 2.5m hight and about 50° angled downwards :)

**Reply by knmcguire** [reply]
Hmm interesting! Perhaps we should try to recreate that here once we have time. The crazyflie should be able to handle 4 basestations still in that configuration. But I think you should also be cover that same area with 3 as well, that might already improve things.

**Reply by krichardsson** [reply]
4 base stations is just about what can be handled in the current implementation, there is very little CPU resources left. Some standard flying should work fine but I can imagine that if you add more load to the system it will start to be overloaded.

A first sign of overload is a warning that the kalman rate is off. The assert that you got in the radio link is usually another indication of overload, it happens when the radio link does not get enough CPU time to handle its duties.

Some options for you are:
* Use 3 base stations instead of 4
* There is a throttling functionality that throws away sweep data from base stations to reduce the load. Unfortunately it is a bit too late in the processing chain to have a big impact but you can try to throttle it down a bit more. You set it with the [lighthouse.lh2maxRate parameter](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#lighthouselh2maxrate)
* Reduce the load of "other stuff". Not sure if you have an app of your own running in the Crazyflie as well, but if you do try to make it as efficient as possible.

**Reply by krichardsson** [reply]
I have added a fix to the firmware that reduces the CPU load when using the lighthouse. With 4 base stations it is freeing up some 15% CPU and 4 base stations should work much better.

The change was implemented in https://github.com/bitcraze/crazyflie-firmware/pull/1308 and is merged to master.
